### DISCUSSION TITLE
Is it possible to override ? (ternary operator)?

### DISCUSSION THREAD

**Discussion by mdmcconnell** [antonmedv/expr#679] [General]
I've been trying to override the ? operator because I would like it to deal with nil values.  Expr accepts my call to expr.Operator("?", "ternWithNil") and complains if I haven't registered ternWithNil function in the environment, and will complain about the signature if the target function doesn't have two (why two?) input parameters and one return value, however it never seems to reach my code.  Is it possible to override ? without building a patcher?

**Comment by mdmcconnell** [comment]
Here is some patcher code that seems to work, but I don't feel confident enough about my understanding of ast to trust it completely.  It would be nice to be able to use the built-in operator override mechanism.

```
type NilMathPatcher struct{}

func (NilMathPatcher) Visit(node *ast.Node) {
	if n, ok := (*node).(*ast.ConditionalNode); ok {
		callNode := &ast.CallNode{
			Callee:    &ast.IdentifierNode{Value: "ternWithNil"},
			Arguments: []ast.Node{n.Cond, n.Exp1, n.Exp2},
		}
		ast.Patch(node, callNode)
	}
}
```

**Comment by antonmedv** [comment]
It is possible to override only binary operators. X?y:z isn’t possible to override. But you can patch it to your needs.
